File: src/go/3_StructuralPatterns/6_Flyweight/textformatting.go
package main

import (
  "fmt"
  "strings"
  "unicode"
)

type FormattedText struct {
  plainText  string
  capitalize []bool
}

func (f *FormattedText) String() string {
  sb := strings.Builder{}
  for i := 0; i < len(f.plainText); i++ {
    c := f.plainText[i]
    if f.capitalize[i] {
      sb.WriteRune(unicode.ToUpper(rune(c)))
    } else {
      sb.WriteRune(rune(c))
    }
  }
  return sb.String()
}

func NewFormattedText(plainText string) *FormattedText {
  return &FormattedText{plainText,
    make([]bool, len(plainText))}
}

func (f *FormattedText) Capitalize(start, end int) {
  for i := start; i <= end; i++ {
    f.capitalize[i] = true
  }
}

type TextRange struct {
  Start, End int
  Capitalize, Bold, Italic bool
}

func (t *TextRange) Covers(position int) bool {
  return position >= t.Start && position <= t.End
}

type BetterFormattedText struct {
  plainText string
  formatting []*TextRange
}

func (b *BetterFormattedText) String() string {
  sb := strings.Builder{}

  for i := 0; i < len(b.plainText); i++ {
    c := b.plainText[i]
    for _, r := range b.formatting {
      if r.Covers(i) && r.Capitalize {
        c = uint8(unicode.ToUpper(rune(c)))
      }
    }
    sb.WriteRune(rune(c))
  }

  return sb.String()
}

func NewBetterFormattedText(plainText string) *BetterFormattedText {
  return &BetterFormattedText{plainText: plainText}
}

func (b *BetterFormattedText) Range(start, end int) *TextRange {
  r := &TextRange{start, end, false, false, false}
  b.formatting = append(b.formatting, r)
  return r
}



func main() {
  text := "This is a brave new world"

  ft := NewFormattedText(text)
  ft.Capitalize(10, 15) // brave
  fmt.Println(ft.String())

  bft := NewBetterFormattedText(text)
  bft.Range(16, 19).Capitalize = true // new
  fmt.Println(bft.String())
}

File: src/csharp/3_StructuralPatterns/6_Flyweight/TextFormatting.cs
using System.Activities.Statements;
using System.Collections.Generic;
using System.Text;
using static System.Console;

namespace DotNetDesignPatternDemos.Structural.Flyweight.TextFormatting
{
  public class FormattedText
  {
    private string plainText;

    public FormattedText(string plainText)
    {
      this.plainText = plainText;
      capitalize = new bool[plainText.Length];
    }

    public void Capitalize(int start, int end)
    {
      for (int i = start; i <= end; ++i)
        capitalize[i] = true;
    }

    private bool[] capitalize;

    public override string ToString()
    {
      var sb = new StringBuilder();
      for (var i = 0; i < plainText.Length; i++)
      {
        var c = plainText[i];
        sb.Append(capitalize[i] ? char.ToUpper(c) : c);
      }
      return sb.ToString();
    }
  }

  public class BetterFormattedText
  {
    private string plainText;
    private List<TextRange> formatting = new List<TextRange>();

    public BetterFormattedText(string plainText)
    {
      this.plainText = plainText;
    }

    public TextRange GetRange(int start, int end)
    {
      var range = new TextRange {Start = start, End = end};
      formatting.Add(range);
      return range;
    }

    public override string ToString()
    {
      var sb = new StringBuilder();

      for (var i = 0; i < plainText.Length; i++)
      {
        var c = plainText[i];
        foreach (var range in formatting)
          if (range.Covers(i) && range.Capitalize)
            c = char.ToUpperInvariant(c);
        sb.Append(c);
      }

      return sb.ToString();
    }

    public class TextRange
    {
      public int Start, End;
      public bool Capitalize, Bold, Italic;

      public bool Covers(int position)
      {
        return position >= Start && position <= End;
      }
    }
  }

  public class Demo
  {
    static void Main(string[] args)
    {
      var ft = new FormattedText("This is a brave new world");
      ft.Capitalize(10, 15);
      WriteLine(ft);

      var bft = new BetterFormattedText("This is a brave new world");
      bft.GetRange(10, 15).Capitalize = true;
      WriteLine(bft);
    }
  }
}